### PR TITLE
use a fixed number of reader/writers for gzip

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

**Comment by mmaxim** [comment]
Actually this won't work...

**Comment by joshblum** [comment]
yeah i don't think we should lock per request, i originally tried a single reader but removed it in https://github.com/keybase/go-framed-msgpack-rpc/pull/163 to avoid locking. what if there was a reader/writer pool?

**Comment by mmaxim** [comment]
Put up something else, not sure if it is better.

File: rpc/compressor.go [hunk 1]
```diff
@@ -11,18 +12,62 @@ type compressor interface {
...
+	return gzipWriter, func() {
+		c.writerLock.Lock()
+		c.gzipWriters = append(c.gzipWriters, gzipWriter)
+		c.writerLock.Unlock()
+	}
+}
+func (c *gzipCompressor) getGzipReader(reader io.Reader) (*gzip.Reader, func(), error) {
+	c.readerLock.Lock()
+	defer c.readerLock.Unlock()
+	var err error
```
Comment (joshblum) [inline rpc/compressor.go]: named return?

File: rpc/compressor.go [hunk 2]
```diff
@@ -11,18 +12,62 @@ type compressor interface {
...
+		c.readerLock.Unlock()
+	}, nil
+}
+
 func (c *gzipCompressor) Compress(data []byte) ([]byte, error) {
 
 	var buf bytes.Buffer
-	writer := gzip.NewWriter(&buf)
+	writer, reclaim := c.getGzipWriter(&buf)
+	defer reclaim()
```
Comment (patrickxb) [inline rpc/compressor.go]: I'm pretty sure writer.Close() will flush before it closes, so the writer.Flush() call is unnecessary.

**Pull request merged**